pkg/cnsmigration/migrator.go

```diff
@@ -0,0 +1,315 @@
+	// find existing CSI based persistent volumes, which same datastore as one mentioned in the source
+	c.createOpenshiftClients()
+
+	err := c.loginToVCenter(ctx)
+	if err != nil {
+		printError("error logging into vcenter: %v", err)
+		return err
+	}
+
+	printGreenInfo("logging successfully to vcenter")
+
+	err = c.getCNSVolumes(ctx, c.sourceDatastore)
+	if err != nil {
+		printError("error listing cns volumes: %v", err)
+		return err
+	}
+	pvList, err := c.clientSet.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	podList, err := c.clientSet.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	c.usedVolumeCache = NewInUseStore(pvList.Items)
+	c.usedVolumeCache.addAllPods(podList.Items)
+	err = c.findAndMigrateCSIVolumes(ctx, volumeFile)
+	c.printSummary()
+	return err
+}
+
+func (c *CNSVolumeMigrator) printSummary() {
+	printGreenInfo("----------- Migration Summary ------------")
+	printGreenInfo("Migrated %d volumes", c.migratedVolumes)
+	printGreenInfo("Failed to migrate %d volumes", c.failedToMigrate)
+	printGreenInfo("Volumes not found %d", c.volumesNotFound)
+}
+
+func (c *CNSVolumeMigrator) findAndMigrateCSIVolumes(ctx context.Context, volumeFile string) error {
+	file, err := os.Open(volumeFile)
+	if err != nil {
+		msg := fmt.Errorf("error opening file %s: %v", volumeFile, err)
+		klog.Error(msg)
+		return msg
+	}
+	defer file.Close()
+
+	fileBytes, err := io.ReadAll(file)
+	if err != nil {
+		msg := fmt.Errorf("error reading file %s: %v", volumeFile, err)
+		klog.Error(msg)
+		return msg
+	}
+	if len(fileBytes) == 0 {
+		msg := fmt.Errorf("file %s has no listed volumes", volumeFile)
+		return msg
+	}
+
+	volumeLines := strings.Split(string(fileBytes), "\n")
+	for _, volumeLine := range volumeLines {
+		pvName := strings.TrimSpace(volumeLine)
+		if pvName == "" {
+			continue
+		}
+		printGreenInfo("Starting migration for pv %s", pvName)
+		pv, err := c.clientSet.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+		if err != nil {
+			msg := fmt.Errorf("error finding pv %s: %v", pvName, err)
+			printErrorObject(msg)
+			return msg
+		}
+
+		csiSource := pv.Spec.CSI
+		if csiSource != nil && csiSource.Driver == vSphereCSIDriverName {
+			if c.checkForDatastore(pv) {
+				err = c.migrateVolume(ctx, csiSource.VolumeHandle)
+				if err != nil {
+					c.failedToMigrate++
+					printError("error migrating volume %s with error: %v", pvName, err)
+				} else {
+					c.migratedVolumes++
+					printGreenInfo("successfully migrated pv %s with volumeID: %s", pvName, csiSource.VolumeHandle)
+				}
+			}
+		} else {
+			printGreenInfo("pv %s is not a vSphere CSI volume", pvName)
+		}
+	}
+	return nil
+}
+
+func (c *CNSVolumeMigrator) migrateVolume(ctx context.Context, volumeID string) error {
+	dtsDatastore, err := c.getDatastore(ctx, c.destinationDatastore)
```

Comment:
The destination datastore could be cached easily, saving one vCenter call per volume.